akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ClusterSharding.scala

```diff
@@ -1778,4 +1813,39 @@ class ShardCoordinator(handOffTimeout: FiniteDuration, shardStartTimeout: Finite
 
   def allocateShardHomes(): Unit = persistentState.unallocatedShards.foreach { self ! GetShardHome(_) }
 
+  def continueGetShardHome(shard: ShardId, region: ActorRef, getShardHomeSender: ActorRef): Unit =
+    if (!rebalanceInProgress.contains(shard)) {
+      persistentState.shards.get(shard) match {
+        case Some(ref) ⇒ getShardHomeSender ! ShardHome(shard, ref)
+        case None ⇒
+          if (persistentState.regions.contains(region)) {
+            persist(ShardHomeAllocated(shard, region)) { evt ⇒
+              persistentState = persistentState.updated(evt)
+              log.debug("Shard [{}] allocated at [{}]", evt.shard, evt.region)
+
+              sendHostShardMsg(evt.shard, evt.region)
+              getShardHomeSender ! ShardHome(evt.shard, evt.region)
+            }
+          } else
+            log.debug("Allocated region {} for shard [{}] is not (any longer) one of the registered regions: {}",
+              region, shard, persistentState)
+      }
+    }
+
+  def continueRebalance(shards: Set[ShardId]): Unit =
+    shards.foreach { shard ⇒
+      if (!rebalanceInProgress(shard)) {
```

Comment:
concurrent rebalance will be handled by this check, i.e. only first one will trigger a rebalance